packages/adapter/src/util.ts

```diff
@@ -0,0 +1,106 @@
+import _ from 'lodash';
+import { u8aToHex, hexToU8a } from '@polkadot/util';
+import { ApiPromise } from '@polkadot/api';
+import { KeyringPair } from '@polkadot/keyring/types';
+import type { SubmittableExtrinsic } from '@polkadot/api/types';
+import type { HexString } from '@polkadot/util/types';
+import { TypeRegistry } from '@polkadot/types';
+import { blake2AsU8a } from '@polkadot/util-crypto';
+import { AccountType, SendExtrinsicResult } from './types';
+
+/**
+ * Send extrinsic
+ * @param api 
+ * @param extrinsic 
+ * @param keyringPair 
+ * @param options 
+ * @returns operation result
+ */
+export const sendExtrinsic = async (
+  api: ApiPromise,
+  extrinsic: SubmittableExtrinsic<'promise'>,
+  keyringPair: KeyringPair,
+  { isSudo = false } = {}
+): Promise<SendExtrinsicResult> => {
+  return new Promise<SendExtrinsicResult>((resolve) => {
+    const newExtrinsic = isSudo ? api.tx.sudo.sudo(extrinsic) : extrinsic;
+    newExtrinsic.signAndSend(keyringPair, { nonce: -1 }, ({ status, events } : any) => {
+      console.log('status.type', status.type);
+      if (status.isInBlock || status.isFinalized) {
+        events
+          // find/filter for failed events
+          .filter(({ event } : any) => api.events.system.ExtrinsicFailed.is(event))
+          // we know that data for system.ExtrinsicFailed is
+          // (DispatchError, DispatchInfo)
+          .forEach(({ event: { data: [error] } }: {event: any}) => {
+            if (error.isModule) {
+              // for module errors, we have the section indexed, lookup
+              const decoded = api.registry.findMetaError(error.asModule);
+              const { docs, method, section } = decoded;
+              console.log(`${section}.${method}: ${docs.join(' ')}`);
+            } else {
+              // Other, CannotLookup, BadOrigin, no extra info
+              console.log(error.toString());
+            }
+          });
+
+        if (status.isFinalized) {
+          resolve({ events, blockHash: status.asFinalized.toString() });
+        }
+      }
+    });
+  });
+};
+
+export const getDerivativeAccountV2 = (api: ApiPromise, accountId: HexString, paraId: number, { locationType = 'XcmV2MultiLocation', network = 'Any' } = {}): HexString => {
+  const account = hexToU8a(accountId).length == 20
+    ? { AccountKey20: { network, key: accountId } }
+    : { AccountId32: { network, id: accountId } };
+
+  const location = {
+    parents: 1,
+    interior: { X2: [{ Parachain: paraId },  account] },
+  };
+  const multilocation = api.createType(locationType, location);
+  const toHash = new Uint8Array([
+    ...new Uint8Array([32]),
+    ...new TextEncoder().encode('multiloc'),
+    ...multilocation.toU8a(),
+  ]);
+
+  return u8aToHex(api.registry.hash(toHash).slice(0, 32));
+};
+
+export const getDerivativeAccountV3 = (accountId: HexString, paraId: number, deriveAccountType: AccountType = AccountType.AccountId32): HexString => {
```

Comment:
Please add comments to all utils functions.

Reply:
Fixed